File: GOLANG/Encapsulamento.go
package main

import (
    "fmt"
    "errors"
)

type ContaBancaria struct {
    titular string
    saldo   float64
}

func NovaConta(titular string, saldoInicial float64) *ContaBancaria {
    return &ContaBancaria{titular, saldoInicial}
}

func (c *ContaBancaria) Depositar(valor float64) {
    if valor > 0 {
        c.saldo += valor
        fmt.Printf("Depositado: R$%.2f. Saldo atual: R$%.2f\n", valor, c.saldo)
    } else {
        fmt.Println("Valor de depósito inválido.")
    }
}

func (c *ContaBancaria) Sacar(valor float64) error {
    if valor <= 0 {
        return errors.New("valor de saque inválido")
    }
    if valor > c.saldo {
        return errors.New("saldo insuficiente")
    }
    c.saldo -= valor
    fmt.Printf("Sacado: R$%.2f. Saldo atual: R$%.2f\n", valor, c.saldo)
    return nil
}

func (c *ContaBancaria) ExibirSaldo() float64 {
    return c.saldo
}

func main() {
    conta := NovaConta("Pedro Souza", 1000.00)

    conta.Depositar(500.00)
    if err := conta.Sacar(200.00); err != nil {
        fmt.Println("Erro:", err)
    }
    if err := conta.Sacar(2000.00); err != nil {
        fmt.Println("Erro:", err)
    }

    fmt.Printf("Saldo atual: R$%.2f\n", conta.ExibirSaldo())
}

File: PYTHON/Encapsulamento.py
class ContaBancaria:
    def __init__(self, titular):
        self.__saldo = 0
        self.titular = titular

    def depositar(self, valor):
        if valor > 0:
            self.__saldo += valor
            print(f'Depositado: R${valor}. Saldo atual: R${self.__saldo}')
        else:
            print('Valor de depósito inválido.')

    def sacar(self, valor):
        if 0 < valor <= self.__saldo:
            self.__saldo -= valor
            print(f'Sacado: R${valor}. Saldo atual: R${self.__saldo}')
        else:
            print('Saldo insuficiente ou valor de saque inválido.')

    def exibir_saldo(self):
        return f'Titular: {self.titular}, Saldo atual: R${self.__saldo}'


conta = ContaBancaria('Pedro Souza')

conta.depositar(1000)
print(conta.exibir_saldo())

conta.sacar(500)
print(conta.exibir_saldo())

conta.sacar(600)
print(conta.exibir_saldo())
